File: assets/ts/breakpoint-interop.ts
const NOTIFY_BREAKPOINT_CHANGED = "OnBreakpointChangedAsync";
const NOTIFY_INITIALIZED = "OnInitializedAsync";

let _ref = null;
let _initialized = false;
let _breakpoint = 0;
let _breakpoints = {};
function getBreakpoint() {
    let width = window.innerWidth;
    if (width >= _breakpoints["Xxl"])
        return 5;
    if (width >= _breakpoints["Xl"])
        return 4;
    else if (width >= _breakpoints["Lg"])
        return 3;
    else if (width >= _breakpoints["Md"])
        return 2;
    else if (width >= _breakpoints["Sm"])
        return 1;
    else //Xs
        return 0;
}

function notifyBreakpointChanged() {
    _ref.invokeMethodAsync(NOTIFY_BREAKPOINT_CHANGED, _breakpoint);
}

function notifyIfBreakpointChanged() {
    let breakpoint = getBreakpoint();
    if(_breakpoint === breakpoint)
        return;
    _breakpoint = breakpoint;
    notifyBreakpointChanged();
}

function notifyInitialized() {
    _ref.invokeMethodAsync(NOTIFY_INITIALIZED, _breakpoint);
}

export function initialize(ref, breakpoints) {
    if(_initialized)
        throw new Error("breakpoint-interop: Already initialized");
    if(!ref)
        throw new Error("breakpoint-interop: ref is not defined");
    _ref = ref;
    if(!breakpoints)
        throw new Error("breakpoint-interop: breakpoints are not defined");
    _breakpoints = breakpoints;
    _initialized = true;
    window.onresize = notifyIfBreakpointChanged;
    notifyIfBreakpointChanged();
    notifyInitialized();
}

File: src/Interop/BreakpointInterop.cs
using Microsoft.JSInterop;

namespace BlazorComps;

public class BreakpointInterop : IAsyncDisposable, IBreakpointInterop
{
    public Func<ValueTask>? OnInitialized { get; set; }
    public Func<ValueTask>? OnBreakpointChanged { get; set; }
    
    private readonly Lazy<Task<IJSObjectReference>> _moduleTask;
    private IReadOnlyDictionary<Breakpoint, int> _breakpoints = new Dictionary<Breakpoint, int>()
    {
        [Breakpoint.Xs] = 0,
        [Breakpoint.Sm] = 576,
        [Breakpoint.Md] = 768,
        [Breakpoint.Lg] = 1024,
        [Breakpoint.Xl] = 1200,
        [Breakpoint.Xxl] = 1400
    };
    
    public bool Initialized { get; private set; }
    public Breakpoint Breakpoint { get; private set; }

    public BreakpointInterop(IJSRuntime jsRuntime)
    {
        _moduleTask = new(
            () => jsRuntime
                .InvokeAsync<IJSObjectReference>(
                    "import",
                    "./_content/BlazorComps/js/breakpoint-interop.js"
                ).AsTask());
    }

    public async ValueTask InitializeAsync()
    {
        var module = await _moduleTask.Value;
        await module.InvokeVoidAsync("initialize", DotNetObjectReference.Create(this), _breakpoints);
    }

    [JSInvokable]
    public async ValueTask OnInitializedAsync()
    {
        if (Initialized)
            throw new InvalidOperationException($"{GetType().Name} has already been initialize");
        Initialized = true;
        if(OnInitialized == null)
            throw new Exception($"{nameof(OnInitialized)} is null.");
        await OnInitialized.Invoke();
    }

    [JSInvokable]
    public async ValueTask OnBreakpointChangedAsync(Breakpoint breakpoint)
    {
        if (Breakpoint == breakpoint)
            return;
        if(OnBreakpointChanged == null)
            throw new Exception($"{nameof(OnBreakpointChanged)} is null.");
        Breakpoint = breakpoint;
        await OnBreakpointChanged.Invoke();
    }

    public async ValueTask DisposeAsync()
    {
        if (_moduleTask.IsValueCreated)
        {
            var module = await _moduleTask.Value;
            await module.DisposeAsync();
        }
    }
}